File: android/app/src/main/java/cz/drabek/chuvicka/parent/Migration.kt
package cz.drabek.chuvicka.parent

import android.content.Context
import cz.drabek.chuvicka.Go2rtc
import cz.drabek.chuvicka.Log
import cz.drabek.chuvicka.Settings
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLDecoder

/**
 * Once, after the update: a phone that reads the camera through go2rtc switches to the camera
 * directly. go2rtc tells the camera's own RTSP URL (with its account) in /api/streams or /api/config.
 * The server settings stay, so the guide can go back to them. The same as ServerMigration on iOS.
 */
object ServerMigration {
    data class Result(
        val host: String, val port: Int, val user: String, val password: String,
        val mainPath: String, val smallPath: String?, val brand: Settings.CameraBrand,
    )

    private const val FLAG = "directMigrated"
    private const val TIMEOUT = 4000

    /**
     * Blocking (up to a few seconds, once): call it on Dispatchers.IO. True when it switched.
     * Runs only for a go2rtc camera and only once; the flag is set also when nothing was found.
     */
    fun run(context: Context): Boolean {
        val prefs = context.getSharedPreferences("migration", Context.MODE_PRIVATE)
        if (prefs.getBoolean(FLAG, false)) return false
        val host = Settings.host.value.trim()
        if (Settings.cameraKind.value != Settings.KIND_GO2RTC || Settings.source.value != Settings.Source.CAMERA || host.isEmpty()) {
            markDone(context)          // A new install or another source: never later.
            return false
        }
        val main = Settings.streamMain.value.trim()
        val small = Settings.streamSmall.value.trim()
        val found = try {
            get(host, "streams")?.let { parseStreams(it, main, small) }
                ?: get(host, "config")?.let { parseConfig(it, main, small) }
        } catch (e: IOException) {
            Log.add("migration: server not reached (${e.message})")
            null
        } catch (e: Exception) {
            Log.add("migration: ${e.javaClass.simpleName}")
            null
        }
        markDone(context)
        if (found == null) {
            Log.add("migration: no direct camera URL in go2rtc, keeping the server")
            return false
        }
        Settings.set(Settings.rtspUrl, "rtspUrl", "rtsp://${found.host}:${found.port}/${found.mainPath}")
        Settings.set(Settings.rtspUrlSmall, "rtspUrlSmall", found.smallPath?.let { "rtsp://${found.host}:${found.port}/$it" } ?: "")
        Settings.set(Settings.rtspUser, "rtspUser", found.user)
        Settings.rtspPassword = found.password
        Settings.set(Settings.rtspBrand, "rtspBrand", found.brand)
        Settings.set(Settings.cameraKind, "cameraKind", Settings.KIND_RTSP)
        Log.add("migrated to the direct camera ${found.host} (${found.brand.title})")
        return true
    }

    /** The guide chose the own server on purpose: never switch it away. */
    fun markDone(context: Context) {
        context.getSharedPreferences("migration", Context.MODE_PRIVATE).edit().putBoolean(FLAG, true).apply()
    }

    /** GET /api/<what>. Null on an answer other than 200; IOException when the server is not there. */
    private fun get(host: String, what: String): String? {
        val c = URL("http://$host:${Go2rtc.API_PORT}/api/$what").openConnection() as HttpURLConnection
        c.connectTimeout = TIMEOUT
        c.readTimeout = TIMEOUT
        try {
            if (c.responseCode != 200) return null
            return c.inputStream.use { it.readBytes() }.toString(Charsets.UTF_8)
        } finally {
            c.disconnect()
        }
    }

    // MARK: Pure: the unit tests run these. Two names, not two parse(): the same Kotlin signature.

    /** /api/streams: {"<name>":{"producers":[{"url":"rtsp://..."}, ...]}, ...}. */
    fun parseStreams(json: String, main: String, small: String): Result? {
        val parsed = try { MiniJson(json).value() } catch (_: Exception) { null }
        val root = parsed as? Map<*, *> ?: return null
        fun url(name: String): String? {
            val producers = (root[name] as? Map<*, *>)?.get("producers") as? List<*> ?: return null
            return producers.firstNotNullOfOrNull { p ->
                ((p as? Map<*, *>)?.get("url") as? String)?.takeIf { it.trim().startsWith("rtsp://", ignoreCase = true) }
            }
        }
        return result(url(main), url(small))
    }

    /** /api/config, the go2rtc YAML: "streams:" then "  <name>:" then "    - rtsp://...". */
    fun parseConfig(yaml: String, main: String, small: String): Result? {
        val lines = yaml.lines()
        val start = lines.indexOfFirst { it.trimEnd() == "streams:" }
        if (start < 0) return null
        // The block ends at the next line that is not indented.
        val block = lines.drop(start + 1).takeWhile { it.isBlank() || it.startsWith(" ") || it.startsWith("\t") || it.startsWith("#") }
        fun url(name: String): String? {
            val at = block.indexOfFirst { it.trim().startsWith("$name:") && indent(it) > 0 }
            if (at < 0) return null
            val own = indent(block[at])
            val inline = unquote(block[at].trim().removePrefix("$name:").substringBefore(" #").trim())
            if (inline.startsWith("rtsp://", ignoreCase = true)) return inline
            for (line in block.drop(at + 1)) {
                val t = line.trim()
                if (t.isEmpty() || t.startsWith("#")) continue
                if (indent(line) < own || (indent(line) == own && !t.startsWith("-"))) break      // The next stream.
                if (!t.startsWith("-")) continue
                val item = unquote(t.removePrefix("-").substringBefore(" #").trim())
                if (item.startsWith("rtsp://", ignoreCase = true)) return item
            }
            return null
        }
        return result(url(main), url(small))
    }

    /** The main stream's URL decides; the sub stream counts only on the same camera. */
    private fun result(mainUrl: String?, smallUrl: String?): Result? {
        val m = split(mainUrl ?: return null) ?: return null
        val s = smallUrl?.let(::split)?.takeIf { it.host == m.host && it.mainPath != m.mainPath }
        return m.copy(smallPath = s?.mainPath)
    }

    /**
     * "rtsp://user:pass@host:port/path?query#fragment" to its parts, the account percent-decoded.
     * By hand, not URI: a go2rtc config may hold a password with a raw "@" or "!".
     */
    private fun split(url: String): Result? {
        val text = url.trim().substringBefore('#')           // go2rtc options ("#backchannel=0").
        if (!text.startsWith("rtsp://", ignoreCase = true)) return null
        val rest = text.substring(7)
        val authority = rest.substringBefore('/')
        val path = rest.substringAfter('/', "")
        val info = authority.substringBeforeLast('@', "")
        val hostPort = authority.substringAfterLast('@')
        val host = hostPort.substringBefore(':')
        if (host.isEmpty()) return null
        val port = hostPort.substringAfter(':', "").toIntOrNull() ?: 554
        val user = decode(info.substringBefore(':'))
        val password = if (':' in info) decode(info.substringAfter(':')) else ""
        return Result(host, port, user, password, path, null, brand(path))
    }

    /** Percent-decoding that keeps "+" (it is a plus in a URL's account, not a space). */
    private fun decode(s: String): String = try {
        URLDecoder.decode(s.replace("+", "%2B"), "UTF-8")
    } catch (_: IllegalArgumentException) {
        s                                                   // A raw "%" in the password.
    }

    fun brand(path: String): Settings.CameraBrand {
        val p = path.lowercase()
        return when {
            p.startsWith("stream1") || p.startsWith("stream2") -> Settings.CameraBrand.TAPO
            p.startsWith("streaming/channels") -> Settings.CameraBrand.HIKVISION
            p.startsWith("cam/realmonitor") -> Settings.CameraBrand.DAHUA
            p.startsWith("h264preview") -> Settings.CameraBrand.REOLINK
            else -> Settings.CameraBrand.OTHER
        }
    }

    private fun indent(line: String) = line.length - line.trimStart().length

    private fun unquote(s: String): String =
        if (s.length >= 2 && (s[0] == '"' || s[0] == '\'') && s.last() == s[0]) s.substring(1, s.length - 1) else s
}

/**
 * A small JSON reader: objects to Map, arrays to List, strings, numbers as Double, true/false, null.
 * org.json is only a stub in the unit tests, so the parse above uses this.
 */
private class MiniJson(private val s: String) {
    private var i = 0

    fun value(): Any? {
        space()
        if (i >= s.length) throw IllegalArgumentException("end")
        return when (val c = s[i]) {
            '{' -> obj()
            '[' -> array()
            '"' -> string()
            't' -> word("true", true)
            'f' -> word("false", false)
            'n' -> word("null", null)
            else -> if (c == '-' || c.isDigit()) number() else throw IllegalArgumentException("at $i")
        }
    }

    private fun obj(): Map<String, Any?> {
        val out = LinkedHashMap<String, Any?>()
        i++
        space()
        if (peek() == '}') { i++; return out }
        while (true) {
            space()
            val key = string()
            space()
            expect(':')
            out[key] = value()
            space()
            when (next()) {
                ',' -> continue
                '}' -> return out
                else -> throw IllegalArgumentException("object at $i")
            }
        }
    }

    private fun array(): List<Any?> {
        val out = ArrayList<Any?>()
        i++
        space()
        if (peek() == ']') { i++; return out }
        while (true) {
            out.add(value())
            space()
            when (next()) {
                ',' -> continue
                ']' -> return out
                else -> throw IllegalArgumentException("array at $i")
            }
        }
    }

    private fun string(): String {
        expect('"')
        val b = StringBuilder()
        while (true) {
            val c = next()
            when (c) {
                '"' -> return b.toString()
                '\\' -> when (val e = next()) {
                    'n' -> b.append('\n')
                    't' -> b.append('\t')
                    'r' -> b.append('\r')
                    'b' -> b.append('\b')
                    'f' -> b.append('\u000C')
                    'u' -> {
                        if (i + 4 > s.length) throw IllegalArgumentException("escape")
                        b.append(s.substring(i, i + 4).toInt(16).toChar())
                        i += 4
                    }
                    else -> b.append(e)                // \" \\ \/
                }
                else -> b.append(c)
            }
        }
    }

    private fun number(): Double {
        val from = i
        while (i < s.length && (s[i].isDigit() || s[i] in "+-.eE")) i++
        return s.substring(from, i).toDouble()
    }

    private fun word(w: String, v: Any?): Any? {
        if (!s.startsWith(w, i)) throw IllegalArgumentException("at $i")
        i += w.length
        return v
    }

    private fun space() { while (i < s.length && s[i].isWhitespace()) i++ }
    private fun peek(): Char = if (i < s.length) s[i] else throw IllegalArgumentException("end")
    private fun next(): Char = if (i < s.length) s[i++] else throw IllegalArgumentException("end")
    private fun expect(c: Char) { if (next() != c) throw IllegalArgumentException("expected $c at $i") }
}

File: Nursery/App/Migration.swift
import Foundation

/// A phone set up with the go2rtc server moves to the camera read directly, once.
///
/// go2rtc knows the camera's own RTSP address with its login (`/api/streams`, else `/api/config`).
/// The app takes it from there, so nobody types anything. The server settings (host, streams,
/// remote host) stay as they were: the guide can go back to the server at any time.
enum ServerMigration {
    struct Result: Equatable {
        let host: String
        let port: Int
        let user: String
        let password: String
        let mainPath: String
        let smallPath: String?
        let brand: CameraBrand
    }

    /// Set after one run that reached the server, with a result or without.
    static let doneKey = "directMigrated"

    private static let session: URLSession = {
        let c = URLSessionConfiguration.ephemeral
        c.timeoutIntervalForRequest = 4
        c.waitsForConnectivity = false
        return URLSession(configuration: c)
    }()

    /// It runs at launch, before the monitor starts. True when the app now reads the camera directly.
    @MainActor static func run(settings: Settings) async -> Bool {
        let defaults = UserDefaults.standard
        guard !MonitorEngine.isDemo, settings.cameraKind == .go2rtc, settings.source == .camera,
              !defaults.bool(forKey: doneKey) else { return false }
        let host = settings.trimmedHost
        guard !host.isEmpty else { return false }
        // Away from home the LAN address does not answer: try again at the next launch,
        // and do not hold up the start for long.
        guard await Reach.canConnect(host: host, port: Go2rtc.apiPort, timeout: 0.8) else { return false }
        let main = settings.streamMain
        let small = settings.streamSmall
        var found: Result?
        if let data = await get(host: host, path: "api/streams") {
            found = parse(streamsJSON: data, main: main, small: small)
        }
        if found == nil, let data = await get(host: host, path: "api/config") {
            found = parse(configYAML: String(decoding: data, as: UTF8.self), main: main, small: small)
        }
        defaults.set(true, forKey: doneKey)
        guard let r = found else {
            Log.shared.add("migration: the server gives no camera address, staying with the server")
            return false
        }
        // A brand only when its path is the one the app would use. Else a full address ("Jiná kamera").
        let known = r.brand != .other && r.mainPath.lowercased() == r.brand.paths.main.lowercased()
        let brand: CameraBrand = known ? r.brand : .other
        settings.rtspBrand = brand
        settings.rtspHost = r.host
        settings.rtspPort = r.port
        settings.rtspUser = r.user
        CameraSecret.password = r.password
        if brand == .other { settings.rtspCustom = "\(r.host):\(r.port)/\(r.mainPath)" }
        settings.cameraKind = .rtsp
        Log.shared.add("migrated to the direct camera \(r.host) (\(brand.rawValue))")
        return true
    }

    private static func get(host: String, path: String) async -> Data? {
        guard let url = URL(string: "http://\(host):\(Go2rtc.apiPort)/\(path)") else { return nil }
        guard let answer = try? await session.data(from: url),
              (answer.1 as? HTTPURLResponse)?.statusCode == 200 else { return nil }
        return answer.0
    }

    // MARK: Parsing (pure, for the tests)

    /// `/api/streams`: `{"nursery":{"producers":[{"url":"rtsp://…"}]}, …}`.
    static func parse(streamsJSON: Data, main: String, small: String) -> Result? {
        guard let object = (try? JSONSerialization.jsonObject(with: streamsJSON)) as? [String: Any] else { return nil }
        func url(of name: String) -> String? {
            guard let stream = object[name] as? [String: Any],
                  let producers = stream["producers"] as? [[String: Any]] else { return nil }
            return producers.compactMap { $0["url"] as? String }.first { $0.lowercased().hasPrefix("rtsp://") }
        }
        return result(main: url(of: main), small: url(of: small))
    }

    /// `/api/config`: the YAML file. Only the `streams:` block is read, and in it the first
    /// `- rtsp://…` item of each stream (or `name: rtsp://…` on one line).
    static func parse(configYAML: String, main: String, small: String) -> Result? {
        let block = streamsBlock(configYAML)
        return result(main: yamlURL(of: main, in: block), small: yamlURL(of: small, in: block))
    }

    /// The lines under the top-level `streams:` key, with no comments and no empty lines.
    private static func streamsBlock(_ yaml: String) -> [String] {
        var inside = false
        var block: [String] = []
        for raw in yaml.components(separatedBy: .newlines) {
            let line = uncommented(raw)
            if line.trimmingCharacters(in: .whitespaces).isEmpty { continue }
            let topLevel = !(line.first?.isWhitespace ?? false)
            if topLevel {
                if inside { break }
                inside = line.trimmingCharacters(in: .whitespaces) == "streams:"
                continue
            }
            if inside { block.append(line) }
        }
        return block
    }

    /// A YAML comment starts with "#" after a space. "rtsp://…#backchannel=0" is not a comment.
    private static func uncommented(_ line: String) -> String {
        if line.trimmingCharacters(in: .whitespaces).hasPrefix("#") { return "" }
        if let r = line.range(of: " #") { return String(line[..<r.lowerBound]) }
        return line
    }

    private static func yamlURL(of name: String, in block: [String]) -> String? {
        guard !name.isEmpty else { return nil }
        var nameIndent: Int?
        for line in block {
            let indent = line.prefix { $0 == " " || $0 == "\t" }.count
            let text = line.trimmingCharacters(in: .whitespaces)
            if let n = nameIndent {
                // The next stream starts: this one has no RTSP source.
                if indent < n || (indent == n && !text.hasPrefix("-")) { return nil }
                if text.hasPrefix("-") {
                    let value = unquoted(String(text.dropFirst()))
                    if value.lowercased().hasPrefix("rtsp://") { return value }
                }
                continue
            }
            for key in [name, "\"\(name)\"", "'\(name)'"] where text.hasPrefix(key + ":") {
                let rest = unquoted(String(text.dropFirst(key.count + 1)))
                if rest.isEmpty {
                    nameIndent = indent
                } else if rest.lowercased().hasPrefix("rtsp://") {
                    return rest
                }
            }
        }
        return nil
    }

    private static func unquoted(_ s: String) -> String {
        let t = s.trimmingCharacters(in: .whitespaces)
        if t.count >= 2, let first = t.first, first == t.last, first == "\"" || first == "'" {
            return String(t.dropFirst().dropLast())
        }
        return t
    }

    private static func result(main: String?, small: String?) -> Result? {
        guard let main, let m = split(main) else { return nil }
        let s = small.flatMap { split($0) }
        // The sub stream counts only when it is the same camera.
        let smallPath = (s?.host == m.host && s?.port == m.port) ? s?.path : nil
        return Result(host: m.host, port: m.port, user: m.user, password: m.password,
                      mainPath: m.path, smallPath: smallPath, brand: brand(path: m.path))
    }

    /// "rtsp://u%40x:p%3Ass@192.168.0.197:554/stream1" to its parts, with the login decoded.
    /// A loopback address is go2rtc reading itself, not a camera.
    private static func split(_ url: String) -> (host: String, port: Int, user: String, password: String, path: String)? {
        guard let u = URLComponents(string: url), u.scheme?.lowercased() == "rtsp",
              let host = u.host, !host.isEmpty,
              host != "localhost", !host.hasPrefix("127."), host != "::1" else { return nil }
        var path = u.percentEncodedPath
        if path.hasPrefix("/") { path.removeFirst() }
        if let query = u.percentEncodedQuery, !query.isEmpty { path += "?" + query }
        let user = u.percentEncodedUser.map { $0.removingPercentEncoding ?? $0 } ?? ""
        let password = u.percentEncodedPassword.map { $0.removingPercentEncoding ?? $0 } ?? ""
        return (host, u.port ?? 554, user, password, path)
    }

    /// The brand from the stream path.
    static func brand(path: String) -> CameraBrand {
        let p = path.lowercased()
        if p.hasPrefix("stream1") || p.hasPrefix("stream2") { return .tapo }
        if p.hasPrefix("streaming/channels") { return .hikvision }
        if p.hasPrefix("cam/realmonitor") { return .dahua }
        if p.hasPrefix("h264preview") { return .reolink }
        return .other
    }
}
